File: whiteout-battle-predictor/frontend/src/lib/combatEngine.ts
/**
 * WHITEOUT SURVIVAL — Frontend Combat Engine
 * Client-side simulation for instant results (<2s).
 * Mirrors backend engine exactly — same formulas.
 */

export interface TroopStats {
  atk: number;
  atkPct: number;
  def: number;
  defPct: number;
  hp: number;
  hpPct: number;
  leth: number;
  lethPct: number;
}

export interface ArmyStats {
  infantry: TroopStats;
  lancer: TroopStats;
  marksman: TroopStats;
}

export interface BattleResult {
  winProbability: number;
  winner: "user" | "enemy";
  roundsFought: number;
  userLossesPct: Record<string, number>;
  enemyLossesPct: Record<string, number>;
  biggestAdvantage: string;
  biggestWeakness: string;
  recommendedFormation: Record<string, number>;
  userDamageScore: number;
  enemyDamageScore: number;
  userDefenseScore: number;
  enemyDefenseScore: number;
}

export interface OptimizerUpgrade {
  troop: string;
  stat: string;
  boostAmount: number;
  efficiency: number;
  label: string;
}

export interface OptimizerTargetResult {
  targetPct: number;
  alreadyAchieved: boolean;
  currentWinPct: number;
  upgrades: OptimizerUpgrade[];
}

// ── CORE FORMULAS ─────────────────────────────────────────────────────────────

export function calcDamage(atk: number, atkPct: number, leth: number, lethPct: number): number {
  return (atk * (1 + atkPct / 100) * leth * (1 + lethPct / 100)) / 100;
}

export function calcDefense(def: number, defPct: number, hp: number, hpPct: number): number {
  return (def * (1 + defPct / 100) * hp * (1 + hpPct / 100)) / 100;
}

function effectiveDamage(raw: number, defense: number): number {
  return Math.max(0, raw - defense * 0.3);
}

// ── TYPE ADVANTAGE ────────────────────────────────────────────────────────────

const TYPE_ADVANTAGE: Record<string, string> = {
  infantry: "lancer",
  lancer: "marksman",
  marksman: "infantry",
};
const ADV_BONUS = 0.15;

// ── BATTLE SIMULATOR ─────────────────────────────────────────────────────────

export function simulateBattle(user: ArmyStats, enemy: ArmyStats, maxRounds = 20): BattleResult {
  const troops = ["infantry", "lancer", "marksman"] as const;

  const getDmg = (army: ArmyStats, t: keyof ArmyStats) =>
    calcDamage(army[t].atk, army[t].atkPct, army[t].leth, army[t].lethPct);
  const getDef = (army: ArmyStats, t: keyof ArmyStats) =>
    calcDefense(army[t].def, army[t].defPct, army[t].hp, army[t].hpPct);

  const userDmg = Object.fromEntries(troops.map(t => [t, getDmg(user, t)]));
  const enemyDmg = Object.fromEntries(troops.map(t => [t, getDmg(enemy, t)]));
  const userDef = Object.fromEntries(troops.map(t => [t, getDef(user, t)]));
  const enemyDef = Object.fromEntries(troops.map(t => [t, getDef(enemy, t)]));

  const userHP: Record<string, number> = { infantry: 100, lancer: 100, marksman: 100 };
  const enemyHP: Record<string, number> = { infantry: 100, lancer: 100, marksman: 100 };

  let round = 0;
  while (round < maxRounds) {
    round++;
    const uAlive = troops.filter(t => userHP[t] > 0);
    const eAlive = troops.filter(t => enemyHP[t] > 0);
    if (!uAlive.length || !eAlive.length) break;

    uAlive.forEach(t => {
      const live = eAlive.filter(x => enemyHP[x] > 0);
      if (!live.length) return;
      const preferred = TYPE_ADVANTAGE[t];
      const target = live.includes(preferred) ? preferred : live[0];
      let dmg = effectiveDamage(userDmg[t], enemyDef[target]);
      if (target === preferred) dmg *= (1 + ADV_BONUS);
      enemyHP[target] = Math.max(0, enemyHP[target] - (dmg * 0.8) / round);
    });

    eAlive.forEach(t => {
      const live = uAlive.filter(x => userHP[x] > 0);
      if (!live.length) return;
      const preferred = TYPE_ADVANTAGE[t];
      const target = live.includes(preferred) ? preferred : live[0];
      let dmg = effectiveDamage(enemyDmg[t], userDef[target]);
      if (target === preferred) dmg *= (1 + ADV_BONUS);
      userHP[target] = Math.max(0, userHP[target] - (dmg * 0.8) / round);
    });
  }

  const uDmgTotal = Object.values(userDmg).reduce((a, b) => a + b, 0);
  const eDmgTotal = Object.values(enemyDmg).reduce((a, b) => a + b, 0);
  const uDefTotal = Object.values(userDef).reduce((a, b) => a + b, 0);
  const eDefTotal = Object.values(enemyDef).reduce((a, b) => a + b, 0);

  const rawRatio = (uDmgTotal + uDefTotal) / Math.max(eDmgTotal + eDefTotal, 0.001);
  const uHpAvg = Object.values(userHP).reduce((a, b) => a + b, 0) / 3;
  const eHpAvg = Object.values(enemyHP).reduce((a, b) => a + b, 0) / 3;

  const winPct = Math.min(97, Math.max(3, Math.round(50 + (rawRatio - 1) * 40 + (uHpAvg - eHpAvg) * 0.3)));

  // Strength analysis
  const statMeta = [
    { base: "atk", pct: "atkPct", label: "ATK" },
    { base: "def", pct: "defPct", label: "DEF" },
    { base: "hp",  pct: "hpPct",  label: "HP" },
    { base: "leth",pct: "lethPct",label: "LETH" },
  ];
  const troopLabels: Record<string, string> = { infantry: "Infantry", lancer: "Lancer", marksman: "Marksman" };

  let biggestAdv = "", advDiff = -Infinity;
  let biggestWeak = "", weakDiff = Infinity;

  troops.forEach(t => {
    statMeta.forEach(({ base, pct, label }) => {
      const uVal = (user[t] as any)[base] * (1 + (user[t] as any)[pct] / 100);
      const eVal = (enemy[t] as any)[base] * (1 + (enemy[t] as any)[pct] / 100);
      const diff = uVal - eVal;
      const name = `${troopLabels[t]} ${label}`;
      if (diff > advDiff) { advDiff = diff; biggestAdv = name; }
      if (diff < weakDiff) { weakDiff = diff; biggestWeak = name; }
    });
  });

  const totalScore = troops.reduce((a, t) => a + userDmg[t] + userDef[t], 0) || 1;
  const formation = Object.fromEntries(
    troops.map(t => [t, Math.round((userDmg[t] + userDef[t]) / totalScore * 100)])
  );
  const remainder = 100 - Object.values(formation).reduce((a, b) => a + b, 0);
  formation.marksman += remainder;

  return {
    winProbability: winPct,
    winner: winPct >= 50 ? "user" : "enemy",
    roundsFought: round,
    userLossesPct: Object.fromEntries(troops.map(t => [t, Math.round(Math.max(0, 100 - userHP[t]))])),
    enemyLossesPct: Object.fromEntries(troops.map(t => [t, Math.round(Math.max(0, 100 - enemyHP[t]))])),
    biggestAdvantage: biggestAdv,
    biggestWeakness: biggestWeak,
    recommendedFormation: formation,
    userDamageScore: uDmgTotal,
    enemyDamageScore: eDmgTotal,
    userDefenseScore: uDefTotal,
    enemyDefenseScore: eDefTotal,
  };
}

// ── REVERSE OPTIMIZER ─────────────────────────────────────────────────────────

export function reverseOptimize(
  user: ArmyStats,
  enemy: ArmyStats,
  targets: number[] = [75, 85, 95]
): OptimizerTargetResult[] {
  const currentPct = simulateBattle(user, enemy).winProbability;

  return targets.map(targetPct => {
    if (currentPct >= targetPct) {
      return { targetPct, alreadyAchieved: true, currentWinPct: currentPct, upgrades: [] };
    }

    const troops = ["infantry", "lancer", "marksman"] as const;
    const statKeys = [
      { key: "atk", step: 200, label: "Attack" },
      { key: "leth", step: 200, label: "Lethality" },
      { key: "def", step: 300, label: "Defense" },
      { key: "hp", step: 300, label: "Health" },
      { key: "atkPct", step: 5, label: "Attack%" },
      { key: "lethPct", step: 5, label: "Lethality%" },
      { key: "defPct", step: 5, label: "Defense%" },
      { key: "hpPct", step: 5, label: "Health%" },
    ];
    const troopLabels: Record<string, string> = { infantry: "Infantry", lancer: "Lancer", marksman: "Marksman" };

    const upgrades: OptimizerUpgrade[] = [];

    troops.forEach(troop => {
      statKeys.forEach(({ key, step, label }) => {
        let boost = 0;
        let reached = false;
        while (boost < 10000 && !reached) {
          boost += step;
          const testUser: ArmyStats = JSON.parse(JSON.stringify(user));
          (testUser[troop] as any)[key] += boost;
          if (simulateBattle(testUser, enemy).winProbability >= targetPct) reached = true;
        }
        if (reached) {
          upgrades.push({
            troop: troopLabels[troop],
            stat: label,
            boostAmount: boost,
            efficiency: (targetPct - currentPct) / boost * 1000,
            label: `+${boost.toLocaleString()} ${troopLabels[troop]} ${label}`,
          });
        }
      });
    });

    upgrades.sort((a, b) => b.efficiency - a.efficiency);
    return { targetPct, alreadyAchieved: false, currentWinPct: currentPct, upgrades: upgrades.slice(0, 8) };
  });
}

File: whiteout-battle-predictor/backend/app/core/combat_engine.py
"""
WHITEOUT SURVIVAL — Battle Predictor
Combat Engine v1.0

Modular, formula-driven simulation engine.
All formulas are isolated here — never in UI or API layers.
"""

from dataclasses import dataclass, field
from typing import Optional
import math


# ── DATA STRUCTURES ──────────────────────────────────────────────────────────

@dataclass
class TroopStats:
    atk: float = 0
    atk_pct: float = 0      # percent bonus (e.g. 35 = 35%)
    def_: float = 0
    def_pct: float = 0
    hp: float = 0
    hp_pct: float = 0
    leth: float = 0
    leth_pct: float = 0


@dataclass
class ArmyStats:
    infantry: TroopStats = field(default_factory=TroopStats)
    lancer: TroopStats = field(default_factory=TroopStats)
    marksman: TroopStats = field(default_factory=TroopStats)


@dataclass
class RoundLog:
    round_num: int
    user_hp: dict     # {infantry: float, lancer: float, marksman: float}
    enemy_hp: dict


@dataclass
class BattleResult:
    win_probability: float          # 0–100
    winner: str                     # "user" | "enemy"
    rounds_fought: int
    user_losses_pct: dict           # {infantry: %, lancer: %, marksman: %}
    enemy_losses_pct: dict
    biggest_advantage: str          # e.g. "Infantry ATK"
    biggest_weakness: str
    recommended_formation: dict     # {infantry: %, lancer: %, marksman: %}
    user_damage_score: float
    enemy_damage_score: float
    user_defense_score: float
    enemy_defense_score: float
    round_log: list[RoundLog]


@dataclass
class OptimizerUpgrade:
    troop: str
    stat: str
    boost_amount: float
    efficiency: float               # win% gain per 1000 stat points
    label: str


@dataclass
class OptimizerResult:
    target_pct: int
    already_achieved: bool
    current_win_pct: float
    upgrades: list[OptimizerUpgrade]


# ── CORE FORMULAS ─────────────────────────────────────────────────────────────

def calc_damage(atk: float, atk_pct: float, leth: float, leth_pct: float) -> float:
    """
    DAMAGE = (Attack × (1 + Attack%) × Lethality × (1 + Lethality%)) / 100
    """
    return (atk * (1 + atk_pct / 100) * leth * (1 + leth_pct / 100)) / 100


def calc_defense(def_: float, def_pct: float, hp: float, hp_pct: float) -> float:
    """
    DEFENSE = (Defense × (1 + Defense%) × Health × (1 + Health%)) / 100
    """
    return (def_ * (1 + def_pct / 100) * hp * (1 + hp_pct / 100)) / 100


def calc_effective_damage(raw_dmg: float, enemy_defense: float) -> float:
    """Damage mitigation via defense (30% absorption rate)."""
    return max(0.0, raw_dmg - enemy_defense * 0.30)


# ── TYPE ADVANTAGE MATRIX ────────────────────────────────────────────────────
# Infantry beats Marksman
# Marksman beats Lancer
# Lancer beats Infantry

TYPE_ADVANTAGE = {
    "infantry": "lancer",
    "lancer": "marksman",
    "marksman": "infantry",
}
ADVANTAGE_BONUS = 0.15  # 15% bonus damage against countered type


# ── SIMULATION ENGINE ─────────────────────────────────────────────────────────

class BattleSimulator:
    """
    Round-based battle simulator.
    Each round: damage exchange → troop reduction → recalculate.
    Stops when one side reaches 0 OR max_rounds reached.
    """

    MAX_ROUNDS = 20
    HP_DECAY_FACTOR = 0.80   # per-round HP drain scaling

    def __init__(self, max_rounds: int = MAX_ROUNDS):
        self.max_rounds = max_rounds

    def simulate(self, user: ArmyStats, enemy: ArmyStats) -> BattleResult:
        troops = ["infantry", "lancer", "marksman"]

        # Pre-calculate static scores
        user_dmg = {t: calc_damage(*self._dmg_args(user, t)) for t in troops}
        enemy_dmg = {t: calc_damage(*self._dmg_args(enemy, t)) for t in troops}
        user_def = {t: calc_defense(*self._def_args(user, t)) for t in troops}
        enemy_def = {t: calc_defense(*self._def_args(enemy, t)) for t in troops}

        # HP pools start at 100 (percentage-based simulation)
        user_hp = {t: 100.0 for t in troops}
        enemy_hp = {t: 100.0 for t in troops}
        round_log = []

        for round_num in range(1, self.max_rounds + 1):
            u_alive = [t for t in troops if user_hp[t] > 0]
            e_alive = [t for t in troops if enemy_hp[t] > 0]

            if not u_alive or not e_alive:
                break

            # User attacks enemy
            for t in u_alive:
                live_targets = [x for x in e_alive if enemy_hp[x] > 0]
                if not live_targets:
                    continue
                preferred = TYPE_ADVANTAGE[t]
                target = preferred if preferred in live_targets else live_targets[0]
                dmg = calc_effective_damage(user_dmg[t], enemy_def[target])
                if target == preferred:
                    dmg *= (1 + ADVANTAGE_BONUS)
                enemy_hp[target] = max(0.0, enemy_hp[target] - dmg * self.HP_DECAY_FACTOR / round_num)

            # Enemy attacks user
            for t in e_alive:
                live_targets = [x for x in u_alive if user_hp[x] > 0]
                if not live_targets:
                    continue
                preferred = TYPE_ADVANTAGE[t]
                target = preferred if preferred in live_targets else live_targets[0]
                dmg = calc_effective_damage(enemy_dmg[t], user_def[target])
                if target == preferred:
                    dmg *= (1 + ADVANTAGE_BONUS)
                user_hp[target] = max(0.0, user_hp[target] - dmg * self.HP_DECAY_FACTOR / round_num)

            round_log.append(RoundLog(
                round_num=round_num,
                user_hp=dict(user_hp),
                enemy_hp=dict(enemy_hp)
            ))

        # ── WIN PROBABILITY CALCULATION ──────────────────────────────────────
        u_dmg_total = sum(user_dmg.values())
        e_dmg_total = sum(enemy_dmg.values())
        u_def_total = sum(user_def.values())
        e_def_total = sum(enemy_def.values())

        raw_ratio = (u_dmg_total + u_def_total) / max(e_dmg_total + e_def_total, 0.001)
        u_hp_avg = sum(user_hp.values()) / 3
        e_hp_avg = sum(enemy_hp.values()) / 3

        win_pct = 50 + (raw_ratio - 1) * 40 + (u_hp_avg - e_hp_avg) * 0.3
        win_pct = max(3.0, min(97.0, round(win_pct, 1)))

        # ── STRENGTH ANALYSIS ────────────────────────────────────────────────
        biggest_adv, adv_diff = "", -float("inf")
        biggest_weak, weak_diff = "", float("inf")

        stat_pairs = [
            ("atk", "atk_pct"),
            ("def_", "def_pct"),
            ("hp", "hp_pct"),
            ("leth", "leth_pct"),
        ]
        stat_labels = {"atk": "ATK", "def_": "DEF", "hp": "HP", "leth": "LETH"}
        troop_labels = {"infantry": "Infantry", "lancer": "Lancer", "marksman": "Marksman"}

        for troop in troops:
            u_t = getattr(user, troop)
            e_t = getattr(enemy, troop)
            for base, pct_key in stat_pairs:
                u_val = getattr(u_t, base) * (1 + getattr(u_t, pct_key) / 100)
                e_val = getattr(e_t, base) * (1 + getattr(e_t, pct_key) / 100)
                diff = u_val - e_val
                label = f"{troop_labels[troop]} {stat_labels[base]}"
                if diff > adv_diff:
                    adv_diff, biggest_adv = diff, label
                if diff < weak_diff:
                    weak_diff, biggest_weak = diff, label

        # ── RECOMMENDED FORMATION ────────────────────────────────────────────
        total_score = sum(user_dmg[t] + user_def[t] for t in troops) or 1
        formation = {t: round((user_dmg[t] + user_def[t]) / total_score * 100) for t in troops}
        remainder = 100 - sum(formation.values())
        formation["marksman"] += remainder  # absorb rounding error

        return BattleResult(
            win_probability=win_pct,
            winner="user" if win_pct >= 50 else "enemy",
            rounds_fought=len(round_log),
            user_losses_pct={t: round(max(0, 100 - user_hp[t])) for t in troops},
            enemy_losses_pct={t: round(max(0, 100 - enemy_hp[t])) for t in troops},
            biggest_advantage=biggest_adv,
            biggest_weakness=biggest_weak,
            recommended_formation=formation,
            user_damage_score=u_dmg_total,
            enemy_damage_score=e_dmg_total,
            user_defense_score=u_def_total,
            enemy_defense_score=e_def_total,
            round_log=round_log,
        )

    def _dmg_args(self, army: ArmyStats, troop: str):
        t = getattr(army, troop)
        return t.atk, t.atk_pct, t.leth, t.leth_pct

    def _def_args(self, army: ArmyStats, troop: str):
        t = getattr(army, troop)
        return t.def_, t.def_pct, t.hp, t.hp_pct


# ── REVERSE OPTIMIZER ─────────────────────────────────────────────────────────

class ReverseOptimizer:
    """
    Solves backwards from target win% to find required stat upgrades.
    Ranks upgrades by efficiency (win% gain per 1000 stat points).
    """

    STEP_SIZES = {
        "atk": 200,
        "leth": 200,
        "def_": 300,
        "hp": 300,
        "atk_pct": 5,
        "def_pct": 5,
        "hp_pct": 5,
        "leth_pct": 5,
    }
    MAX_BOOST = 10000
    TARGETS = [75, 85, 95]

    def __init__(self):
        self.simulator = BattleSimulator()

    def optimize(self, user: ArmyStats, enemy: ArmyStats,
                 targets: list[int] = None) -> list[OptimizerResult]:
        if targets is None:
            targets = self.TARGETS

        current_result = self.simulator.simulate(user, enemy)
        current_pct = current_result.win_probability
        results = []

        for target_pct in targets:
            if current_pct >= target_pct:
                results.append(OptimizerResult(
                    target_pct=target_pct,
                    already_achieved=True,
                    current_win_pct=current_pct,
                    upgrades=[]
                ))
                continue

            upgrades = []
            troop_names = ["infantry", "lancer", "marksman"]

            for troop in troop_names:
                for stat_key, step in self.STEP_SIZES.items():
                    boost = 0
                    reached = False

                    while boost < self.MAX_BOOST and not reached:
                        boost += step
                        import copy
                        test_user = copy.deepcopy(user)
                        troop_obj = getattr(test_user, troop)
                        current_val = getattr(troop_obj, stat_key)
                        setattr(troop_obj, stat_key, current_val + boost)

                        test_result = self.simulator.simulate(test_user, enemy)
                        if test_result.win_probability >= target_pct:
                            reached = True

                    if reached and boost > 0:
                        efficiency = (target_pct - current_pct) / boost * 1000
                        stat_display = {
                            "atk": "Attack", "leth": "Lethality",
                            "def_": "Defense", "hp": "Health",
                            "atk_pct": "Attack%", "leth_pct": "Lethality%",
                            "def_pct": "Defense%", "hp_pct": "Health%"
                        }
                        troop_display = {
                            "infantry": "Infantry",
                            "lancer": "Lancer",
                            "marksman": "Marksman"
                        }
                        upgrades.append(OptimizerUpgrade(
                            troop=troop_display[troop],
                            stat=stat_display[stat_key],
                            boost_amount=boost,
                            efficiency=round(efficiency, 2),
                            label=f"+{int(boost):,} {troop_display[troop]} {stat_display[stat_key]}"
                        ))

            # Rank by efficiency (highest first)
            upgrades.sort(key=lambda u: u.efficiency, reverse=True)

            results.append(OptimizerResult(
                target_pct=target_pct,
                already_achieved=False,
                current_win_pct=current_pct,
                upgrades=upgrades[:8]  # Top 8 recommendations
            ))

        return results
